src/Agent.Listener/MessageListener.cs

```diff
@@ -143,30 +163,66 @@ public sealed class MessageListener : AgentService, IMessageListener
                     {
                         _lastMessageId = message.MessageId;
                     }
+
+                    if (consecutiveErrors > 1) //print the message once only if there was an error
+                    {
+                        _term.WriteLine(StringUtil.Loc("QueueRestored"));
+                    }
+
+                    consecutiveErrors = 0;
                 }
-                catch (TimeoutException)
+                catch (TimeoutException ex)
                 {
                     Trace.Verbose($"{nameof(TimeoutException)} received.");
-                }
-                catch (TaskCanceledException)
-                {
-                    Trace.Verbose($"{nameof(TaskCanceledException)} received.");
+                    //retry after a delay
+                    errorMessage = ex.Message;
                 }
                 catch (TaskAgentSessionExpiredException)
                 {
                     Trace.Verbose($"{nameof(TaskAgentSessionExpiredException)} received.");
-                    // TODO: Throw a specific exception so the caller can control the flow appropriately.
-                    throw;
+                    if (!await CreateSessionAsync(token))
```

Comment:
This was the easiest place to retry. The other option was to throw to the parent, catch it there and retry. I'm not sure when this exception is thrown. The old C# agent is handling it by restarting the message queue.